llama_index/chat_engine/condense_question.py

```diff
@@ -166,6 +214,54 @@ async def achat(
                 ChatMessage(role=MessageRole.ASSISTANT, content=str(response)),
             ]
         )
+
+        return response
+
+    async def astream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> STREAMING_CHAT_RESPONSE_TYPE:
+        chat_history = chat_history or self._chat_history
+
+        # Generate standalone question from conversation context and last message
+        condensed_question = await self._acondense_question(chat_history, message)
+
+        log_str = f"Querying with: {condensed_question}"
+        logger.info(log_str)
+        if self._verbose:
+            print(log_str)
+
+        # Query with standalone question
+        response = await self._query_engine.aquery(condensed_question)
+
+        # Record response
+        if (
+            isinstance(response, StreamingResponse)
+            and response.response_gen is not None
+        ):
+
+            def gen(
+                chat_history: List[ChatMessage], response_gen: TokenGen
+            ) -> TokenGen:
+                response_str = ""
+                for token in response_gen:
+                    response_str += token
+                    yield token
+
+                chat_history.extend(
+                    [
+                        ChatMessage(role=MessageRole.USER, content=message),
+                        ChatMessage(role=MessageRole.ASSISTANT, content=response_str),
+                    ]
+                )
+
+            # override the generator to include writing to chat history
+            response = StreamingResponse(
+                gen(chat_history, response.response_gen),
+                source_nodes=response.source_nodes,
+                metadata=response.metadata,
+            )
```

Comment:
Is this an exact duplica of the logic above? I couldn't spot differences. If so, could we pull it out into a utils function?

Reply:
Yea good point, this can be combined